wgpu/src/backend/direct.rs

```diff
@@ -1264,27 +1269,35 @@ impl crate::Context for Context {
                             &remaining_arrayed_buffer_bindings[array.len()..];
                         bm::BindingResource::BufferArray(Borrowed(slice))
                     }
-                    BindingResource::Sampler(sampler) => bm::BindingResource::Sampler(sampler.id),
+                    BindingResource::Sampler(sampler) => {
+                        bm::BindingResource::Sampler(*sampler.id.downcast_id::<Self::SamplerId>())
+                    }
                     BindingResource::SamplerArray(array) => {
-                        let slice = &remaining_arrayed_samplers[..array.len()];
+                        let samplers = remaining_arrayed_samplers[..array.len()]
+                            .iter()
+                            .map(|id| *id.downcast_id::<Self::SamplerId>())
+                            .collect::<Vec<_>>();
                         remaining_arrayed_samplers = &remaining_arrayed_samplers[array.len()..];
-                        bm::BindingResource::SamplerArray(Borrowed(slice))
-                    }
-                    BindingResource::TextureView(texture_view) => {
-                        bm::BindingResource::TextureView(texture_view.id)
+                        bm::BindingResource::SamplerArray(Owned(samplers))
                     }
+                    BindingResource::TextureView(texture_view) => bm::BindingResource::TextureView(
+                        *texture_view.id.downcast_id::<Self::TextureViewId>(),
```

Comment:
Why can't these types be inferred?

Reply:
I believe these can be inferred, I didn't figure it out until later